bonsai/UseCases/BudgetGateway.swift

```diff
@@ -0,0 +1,83 @@
+//
+//  BudgetGateway.swift
+//  bonsai
+//
+//  Created by hoang on 21.12.2021.
+//
+
+import Foundation
+import CoreData
+
+public protocol BudgetProtocol {
+   var budgetID: String { get }
+   var name: String { get }
+   var amount: Double { get }
+   var period: Int64 { get } // days ??
+}
+
+public final class BudgetGateway {
+
+   let mainContext: NSManagedObjectContext
+
+   init(mainContext: NSManagedObjectContext = DataController.sharedInstance.mainContext) {
+      self.mainContext = mainContext
+   }
+
+   @discardableResult
+   func create(name: String, amount: Double, period: Int64) throws -> Budget {
+      let budget = Budget(
+         context: mainContext,
+         name: name,
+         amount: amount,
+         period: period)
+      try mainContext.save()
+      return budget
+   }
+
+   func update(budget: Budget) throws {
+      try mainContext.save()
+   }
+
+   func delete(budget: Budget) throws {
+      mainContext.delete(budget)
+      try mainContext.save()
+   }
+
+   func getBudget(name: String) throws -> BudgetProtocol {
+      let request: NSFetchRequest<Budget> = Budget.fetchRequest()
+      request.fetchLimit = 1
+      request.predicate = NSPredicate(format: "name == %@", name)
+      guard let budget = try mainContext.fetch(request).first else {
+         throw BudgetDoesntExist()
+      }
+      return budget
+   }
+
+   func getAllBudgets() throws -> [BudgetProtocol] {
+      let request: NSFetchRequest<Budget> = Budget.fetchRequest()
+      let budgets = try mainContext.fetch(request)
+      return budgets
+   }
+
+   func spentCapacity(budgetName: String) throws -> Double {
```

Comment:
I don't really like doing the calculations in this class.
Calculation is only valid if we have a Budget, and we don't need anything else to perform the calculation. Therefore I think the best place to place calculations would be the Budget itself (calculated property in Budget or in an extension)

Also the Single Responsability Principle is broke in the class and in the methods, as we both GETTING/SAVING from storage and performing calculations